sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java

```diff
@@ -0,0 +1,250 @@
+        .apply(HCatalogIO.write()
+            .withConfigProperties(getConfigProperties())
+            .withTable("myowntable"));
+    pipeline.run();
+  }
+
+  /**
+   * Test of Write without specifying a table.
+   */
+  @Test
+  public void testWriteFailureValidationTable() throws Exception {
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage(containsString("table"));
+    HCatalogIO.write().withConfigProperties(getConfigProperties()).validate(null);
+  }
+
+  /**
+   * Test of Write without specifying configuration properties.
+   */
+  @Test
+  public void testWriteFailureValidationConfigProp() throws Exception {
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage(containsString("configProperties"));
+    HCatalogIO.write().withTable("myowntable").validate(null);
+  }
+
+  /**
+   * Test of Read from a non-existent table.
+   */
+  @Test
+  public void testReadFailureTableDoesNotExist() throws Exception {
+    pipeline.apply(HCatalogIO.read()
+        .withConfigProperties(getConfigProperties())
+        .withTable("myowntable"));
+    thrown.expectCause(isA(NoSuchObjectException.class));
+    pipeline.run();
+  }
+
+  /**
+   * Test of Read without specifying configuration properties.
+   */
+  @Test
+  public void testReadFailureValidationConfig() throws Exception {
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage(containsString("configProperties"));
+    HCatalogIO.read().withTable("myowntable").validate(null);
+  }
+
+  /**
+   * Test of Read without specifying a table.
+   */
+  @Test
+  public void testReadFailureValidationTable() throws Exception {
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage(containsString("table"));
+    HCatalogIO.read().withConfigProperties(getConfigProperties()).validate(null);
+  }
+
+  /**
+   * Test of Read using SourceTestUtils.readFromSource(..).
+   */
+  @Test
+  @TestDataSetupReqd
+  public void testReadFromSource() throws Exception {
+    List<BoundedHCatalogSource> sourceList = getSourceList();
+    List<String> recordsList = new ArrayList<String>();
+    for (int i = 0; i < sourceList.size(); i++) {
+      List<DefaultHCatRecord> hcatRecordsList = SourceTestUtils.readFromSource(sourceList.get(i),
+        PipelineOptionsFactory.create());
+      for (HCatRecord record : hcatRecordsList) {
+        recordsList.add(record.get(0).toString());
+      }
+    }
+    Assert.assertThat(getExpectedRecords(TEST_RECORDS_COUNT),
+        Matchers.containsInAnyOrder(recordsList.toArray()));
+  }
+
+  /**
+   * Test of Read using SourceTestUtils.assertSourcesEqualReferenceSource(..).
+   */
+  @Test
+  @TestDataSetupReqd
+  public void testSourcesEqualReferenceSource() throws Exception {
+    List<BoundedHCatalogSource> sourceList = getSourceList();
+    for (int i = 0; i < sourceList.size(); i++) {
+      SourceTestUtils.assertSourcesEqualReferenceSource(sourceList.get(i),
```

Comment:
`assertSourcesEqualReferenceSource` would be more appropriate to call on a top-level, unsplit source - but here you're calling it on results of splitting (produced in `getSourceList`). It's going to also be much faster because you'll call this function only once.